virttest/utils_misc.py

```diff
@@ -329,60 +335,32 @@ def find_command(cmd):
     return utils_path.find_command(cmd)
 
 
-def pid_exists(pid):
-    """
-    Return True if a given PID exists.
-
-    :param pid: Process ID number.
-    """
-    try:
-        os.kill(pid, 0)
-        return True
-    except Exception:
-        return False
-
-
-def safe_kill(pid, signal):
-    """
-    Attempt to send a signal to a given process that may or may not exist.
-
-    :param signal: Signal number.
-    """
-    try:
-        os.kill(pid, signal)
-        return True
-    except Exception:
-        return False
-
-
-def kill_process_tree(pid, sig=signal.SIGKILL):
+def kill_process_tree(pid, sig=signal.SIGKILL, send_sigcont=True, timeout=0):
     """Signal a process and all of its children.
 
     If the process does not exist -- return.
 
     :param pid: The pid of the process to signal.
     :param sig: The signal to send to the processes.
+    :param send_sigcont: Send SIGCONT to allow destroying stopped processes
+    :param timeout: How long to wait for the pid(s) to die
+                    (negative=infinity, 0=don't wait,
+                    positive=number_of_seconds)
     """
-    if not safe_kill(pid, signal.SIGSTOP):
-        return
-    children = decode_to_text(process.system_output("ps --ppid=%d -o pid=" % pid,
-                                                    shell=True, ignore_status=True)).split()
-    for child in children:
-        kill_process_tree(int(child), sig)
-    safe_kill(pid, sig)
-    safe_kill(pid, signal.SIGCONT)
-
-
-def kill_process_by_pattern(pattern):
-    """Send SIGTERM signal to a process with matched pattern.
-    :param pattern: normally only matched against the process name
-    """
-    cmd = "pkill -f %s" % pattern
-    result = process.run(cmd, ignore_status=True)
-    if result.exit_status:
-        logging.error("Failed to run '%s': %s", cmd, result)
-    else:
-        logging.info("Succeed to run '%s'.", cmd)
+    try:
+        return _kill_process_tree(pid, sig, send_sigcont, timeout)
+    except TypeError:
+        logging.warning("Trying to kill_process_tree with timeout but running"
+                        " old Avocado without it's support. Sleeping for 10s "
+                        "instead.")
+        # Depending on the Avocado version this can either return None or
+        # list of killed pids.
+        ret = _kill_process_tree(pid, sig, send_sigcont)    # pylint: disable=E1128
+        if timeout != 0:
+            time.sleep(10)
+            if not pid_exists(pid):
```

Comment:
Should be `if pid_exists(pid):`, right?

Reply:
sure, sorry for not testing this. Please let me know what you think about the 10s vs timeout and I'll update this as well.